### PR TITLE
[PnP] Add support for convention-based command operations

### PR DESCRIPTION
**Pull request by abhipsaMisra**
This PR addresses support for the command APIs.
The E2E tests will be added in a subsequent PR.

Source: https://github.com/Azure/azure-iot-sdk-csharp/tree/previews/pnpImplDev

### CODE REVIEW

File: iothub/device/src/CommandRequest.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        /// The command name.
+        /// </summary>
+        public string CommandName { get; private set; }
+
+        /// <summary>
+        /// The command request data.
+        /// </summary>
+        /// <typeparam name="T">The type to cast the command request data to.</typeparam>
+        /// <returns>The command request data.</returns>
+        public T GetData<T>()
```
Comment (jamdavi) [inline iothub/device/src/CommandRequest.cs:40]: TryGetData<T> ?

Should we also add TryGetValue<T> to match the payload collection?

File: iothub/device/src/InternalClient.ConventionBasedOperations.cs [hunk 2]
```diff
@@ -32,5 +33,33 @@ internal Task SendTelemetryAsync(TelemetryMessage telemetryMessage, Cancellation
...
             return SendEventAsync(telemetryMessage, cancellationToken);
         }
+
+        internal Task SubscribeToCommandsAsync(Func<CommandRequest, object, Task<CommandResponse>> callback, object userContext, CancellationToken cancellationToken)
+        {
+            // Subscribe to methods default handler internally and use the callback received internally to invoke the user supplied command callback.
+            var methodDefaultCallback = new MethodCallback(async (methodRequest, userContext) =>
+            {
+                CommandRequest commandRequest;
+                if (methodRequest.Name.Contains(ConventionBasedConstants.ComponentLevelCommandSeparator))
```
Comment (vinagesh) [inline iothub/device/src/InternalClient.ConventionBasedOperations.cs:43]: any chance methodRequest or methodRequest.Name can be null?

Comment (abhipsaMisra) [reply iothub/device/src/InternalClient.ConventionBasedOperations.cs:43]: there won't be, a method request will always be accompanied by method name; but no harm in adding a null check here

File: iothub/device/src/CommandRequest.cs (hunk 1 above)
Comment (abhipsaMisra) [reply iothub/device/src/CommandRequest.cs:40]: done

**Review by vinagesh — APPROVED** [review]

**Pull request merged**